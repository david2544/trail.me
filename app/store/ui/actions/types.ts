import { ActionType } from 'typesafe-actions';
import * as actions from '@store/ui/actions';

/* --- STATE --- */
interface UIState {
  readonly locale: string;
}

/* --- ACTIONS --- */
type UIActions = ActionType<typeof actions>;

/* --- EXPORTS --- */
export { UIActions, UIState };
