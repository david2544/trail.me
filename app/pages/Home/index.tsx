import React, { useState } from 'react';
import useToggleLocale from '@app/hooks/useToggleLocale';
import styles from './styles.module.scss';

const Home: React.FC = () => {
  const [hovered, setHovered] = useState('');
  const { locale, toggleLocale } = useToggleLocale();

  return (
    <div className={styles.home}>
      <div data-language={locale} data-lang-hover={hovered}>
        <div className={styles.langSelection} />
        <a
          role="button"
          tabIndex={0}
          onClick={() => toggleLocale('de')}
          onMouseEnter={() => setHovered('de')}
          onMouseLeave={() => setHovered(locale)}
          data-lang="de"
          data-active={locale === 'de'}
        >
          de
        </a>
        <a
          role="button"
          tabIndex={0}
          onClick={() => toggleLocale('en')}
          onMouseEnter={() => setHovered('en')}
          onMouseLeave={() => setHovered(locale)}
          data-lang="en"
          data-active={locale === 'en'}
        >
          en
        </a>
      </div>
    </div>
  );
};

export default Home;
