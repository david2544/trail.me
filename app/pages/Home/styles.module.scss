div[data-language] {
  margin: 20px 10px;
  box-sizing: content-box;
  position: relative;
  color: inherit;
  width: 60px;
  display: inline-block;
  white-space: nowrap;

  &:after {
    content: '';
    clear: both;
  }

  .langSelection {
    display: inline-block;
    border-bottom: 1px solid black;
    position: absolute;
    left: 0;
    width: 50%;
    height: 100%;
    transition: transform 100ms ease-in-out;
    pointer-events: none;
  }

  &[data-language='de'],
  &[data-language='en'][data-lang-hover='de'] {
    .langSelection {
      transform: translateX(0px);
    }
  }

  &[data-language='en'],
  &[data-language='de'][data-lang-hover='en'] {
    .langSelection {
      transform: translateX(100%);
      pointer-events: none;
    }
  }

  > a[data-lang] {
    padding: 0 5px;
    z-index: 2;
    position: relative;
    float: left;
    width: 30px;
    text-align: center;
    color: inherit;

    &,
    &:hover,
    &:focus,
    &:active {
      text-decoration: none;
    }

    &[data-active='true'] {
      pointer-events: none;
    }
  }

  @media (max-width: 767) {
    margin: 0;
    width: 100%;
    order: 3;

    .langSelection {
      display: none;
    }

    > a[data-lang] {
      width: 50%;

      &[data-active='true'] {
        text-decoration: underline;
      }
    }
  }
}
