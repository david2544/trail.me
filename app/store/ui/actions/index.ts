import ActionTypes from '@store/actionTypes';
import { action } from 'typesafe-actions';

const changeTheme = (darkMode: boolean) => action(ActionTypes.CHANGE_DARK_MODE, darkMode);

// eslint-disable-next-line import/prefer-default-export
export { changeTheme };
