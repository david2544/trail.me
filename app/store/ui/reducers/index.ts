import ActionTypes from '@store/actionTypes';
import { UIActions, UIState } from '@store/ui/actions/types';

const DEFAULT_LOCALE = 'en';

export const initialState: UIState = {
  locale: DEFAULT_LOCALE,
};

export default (state: UIState = initialState, action: UIActions): UIState => {
  switch (action.type) {
    case ActionTypes.CHANGE_LOCALE:
      return {
        locale: action.payload,
      };
    default:
      return state;
  }
};
