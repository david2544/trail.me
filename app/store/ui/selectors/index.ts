import { ApplicationRootState } from '@store/IStore';
import { initialState } from '../reducers';

/**
 * Direct selector to the languageToggle state domain
 */
const uiState = (state: ApplicationRootState) => state.ui || initialState;

export { uiState };
