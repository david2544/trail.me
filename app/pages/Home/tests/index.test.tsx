import React from 'react';
import * as Redux from 'react-redux';
import { shallow } from 'enzyme';
import * as actions from '@store/ui/actions';
import Home from '..';

jest.mock('store/ui/actions');

describe('<Home />', () => {
  let useSelectorSpy: jest.SpyInstance;
  let useDispatchSpy: jest.SpyInstance;

  beforeEach(() => {
    useSelectorSpy = jest.spyOn(Redux, 'useSelector');
    useDispatchSpy = jest.spyOn(Redux, 'useDispatch');
    useSelectorSpy.mockReturnValue(jest.fn());
    useDispatchSpy.mockReturnValue(jest.fn());
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should match snapshot', () => {
    const wrapper = shallow(<Home />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should call changeLocale action on "de" locale link click', () => {
    const wrapper = shallow(<Home />);
    // link for "de" locale
    const link = wrapper.find('a').first();
    const newLocale = 'de';

    link.simulate('click');
    expect(actions.changeLocale).toHaveBeenCalledWith(newLocale);
  });

  it('should call changeLocale action on "en" locale link click', () => {
    const wrapper = shallow(<Home />);
    // link for "en" locale
    const link = wrapper.find('a').at(1);
    const newLocale = 'en';

    link.simulate('click');
    expect(actions.changeLocale).toHaveBeenCalledWith(newLocale);
  });

  it('should change the data-lang-hover attribute to the "de" locale', () => {
    const wrapper = shallow(<Home />);
    // link for "de" locale
    const link = wrapper.find('a').first();

    link.simulate('mouseEnter');
    expect(wrapper.find('[data-lang-hover="de"]').exists()).toBeTruthy();
  });

  it('should change the data-lang-hover attribute to the "en" locale', () => {
    const wrapper = shallow(<Home />);
    // link for "en" locale
    const link = wrapper.find('a').at(1);

    link.simulate('mouseEnter');
    expect(wrapper.find('[data-lang-hover="en"]').exists()).toBeTruthy();
  });

  it('should change the data-lang-hover attribute to the default locale on mouseLeave for "en" link', () => {
    // mock default value from store
    useSelectorSpy.mockReturnValue({ locale: 'someDefaultValue' });
    const wrapper = shallow(<Home />);
    // link for "de" locale
    const link = wrapper.find('a').first();

    link.simulate('mouseLeave');
    expect(wrapper.find('[data-lang-hover="someDefaultValue"]').exists()).toBeTruthy();
  });

  it('should change the data-lang-hover attribute to the default locale on mouseLeave for "de" link', () => {
    // mock default value from store
    useSelectorSpy.mockReturnValue({ locale: 'someDefaultValue' });
    const wrapper = shallow(<Home />);
    // link for "en" locale
    const link = wrapper.find('a').at(1);

    link.simulate('mouseLeave');
    expect(wrapper.find('[data-lang-hover="someDefaultValue"]').exists()).toBeTruthy();
  });
});
