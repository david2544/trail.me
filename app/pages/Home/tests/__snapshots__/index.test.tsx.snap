// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> should match snapshot 1`] = `
<div
  className="home"
>
  <div
    data-lang-hover=""
  >
    <div
      className="langSelection"
    />
    <a
      data-active={false}
      data-lang="de"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      role="button"
      tabIndex={0}
    >
      de
    </a>
    <a
      data-active={false}
      data-lang="en"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      role="button"
      tabIndex={0}
    >
      en
    </a>
  </div>
</div>
`;
